File: src/unique/unique.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/schollz/progressbar/v3"
	"github.com/shirou/gopsutil/cpu"
	"github.com/shirou/gopsutil/mem"
	"github.com/xuri/excelize/v2"
)

func main() {
	// Caminhos
	userprofile, _ := os.UserHomeDir()
	filesPath := filepath.Join(userprofile, "Documentos", "PythonVsGo_Dataset", "documentos_excel")
	logPath := filepath.Join(userprofile, "Área de Trabalho", "Projetos", "PythonVsGo-FileManipulation", "PythonVsGo-FileManipulation", "logs", "unique_go_log.txt")

	// Monitoramento de desempenho (início)
	startTime := time.Now()
	cpuUsageBefore, _ := cpu.Percent(0, false)
	memUsageBefore, _ := mem.VirtualMemory()

	// Processamento dos arquivos
	uniqueLinesCount, err := countUniqueLines(filesPath)
	if err != nil {
		fmt.Println("Erro ao processar arquivos:", err)
		return
	}

	// Monitoramento de desempenho (fim)
	endTime := time.Now()
	elapsedTime := endTime.Sub(startTime)
	cpuUsageAfter, _ := cpu.Percent(0, false)
	memUsageAfter, _ := mem.VirtualMemory()

	// Salvando em um arquivo de log
	logData := fmt.Sprintf("Time elapsed: %v\nCPU Usage Before: %v%%, After: %v%%\nMemory Usage Before: %v%%, After: %v%%\n",
		elapsedTime, cpuUsageBefore, cpuUsageAfter, memUsageBefore.UsedPercent, memUsageAfter.UsedPercent)

	for file, count := range uniqueLinesCount {
		logData += fmt.Sprintf("%s: %d linhas únicas\n", file, count)
	}

	if err := os.WriteFile(logPath, []byte(logData), 0644); err != nil {
		fmt.Println("Erro ao escrever no arquivo de log:", err)
		return
	}

	fmt.Println("Process finished!")
}

// countUniqueLines conta linhas únicas em arquivos Excel no diretório especificado
func countUniqueLines(dir string) (map[string]int, error) {
	uniqueLines := make(map[string]int)
	files, err := filepath.Glob(filepath.Join(dir, "*.xlsx"))
	if err != nil {
		return nil, err
	}

	bar := progressbar.Default(int64(len(files)), "Processando arquivos")

	for _, file := range files {
		f, err := excelize.OpenFile(file)
		if err != nil {
			return nil, err
		}

		rows, err := f.GetRows(f.GetSheetName(1))
		if err != nil {
			return nil, err
		}

		uniqueRowCount := countUniqueRows(rows)
		uniqueLines[filepath.Base(file)] = uniqueRowCount

		bar.Add(1)
	}

	return uniqueLines, nil
}

// countUniqueRows conta as linhas únicas em uma lista de linhas
func countUniqueRows(rows [][]string) int {
	uniqueRows := make(map[string]bool)
	for _, row := range rows {
		rowStr := fmt.Sprintf("%v", row)
		uniqueRows[rowStr] = true
	}
	return len(uniqueRows)
}

File: src/unique/unique.py
import os
import time
import psutil
from openpyxl import load_workbook
from tqdm import tqdm

def main():
    # Caminhos
    userprofile = os.environ['USERPROFILE']
    filesPath = os.path.join(userprofile, "Documentos", "PythonVsGo_Dataset", "documentos_excel")
    logPath = os.path.join(userprofile, "Área de Trabalho", "Projetos", "PythonVsGo-FileManipulation", "PythonVsGo-FileManipulation", "logs", "unique_python_log.txt")

    # Monitoramento de desempenho (início)
    startTime = time.time()
    cpuUsageBefore = psutil.cpu_percent()
    memUsageBefore = psutil.virtual_memory().percent

    # Processamento dos arquivos
    uniqueLinesCount, err = countUniqueLines(filesPath)
    if err is not None:
        print("Erro ao processar arquivos:", err)
        return

    # Monitoramento de desempenho (fim)
    endTime = time.time()
    elapsedTime = endTime - startTime
    cpuUsageAfter = psutil.cpu_percent()
    memUsageAfter = psutil.virtual_memory().percent

    # Salvando em um arquivo de log
    logData = f"Time elapsed: {elapsedTime} seconds\n"
    logData += f"CPU Usage Before: {cpuUsageBefore}%, After: {cpuUsageAfter}%\n"
    logData += f"Memory Usage Before: {memUsageBefore}%, After: {memUsageAfter}%\n"

    for file, count in uniqueLinesCount.items():
        logData += f"{file}: {count} linhas únicas\n"

    try:
        with open(logPath, "w") as logFile:
            logFile.write(logData)
    except Exception as e:
        print("Erro ao escrever no arquivo de log:", e)
        return

    print("Process finished!")

# countUniqueLines conta linhas únicas em arquivos Excel no diretório especificado
def countUniqueLines(dir):
    uniqueLines = {}
    files = [f for f in os.listdir(dir) if f.endswith(".xlsx")]
    with tqdm(total=len(files), desc="Processando arquivos") as pbar:
        for file in files:
            file_path = os.path.join(dir, file)
            uniqueRowCount, err = countUniqueRows(file_path)
            if err is None:
                uniqueLines[file] = uniqueRowCount
            pbar.update(1)
    return uniqueLines, None

# countUniqueRows conta as linhas únicas em um arquivo Excel
def countUniqueRows(file_path):
    try:
        workbook = load_workbook(filename=file_path, read_only=True)
        sheet = workbook.active
        uniqueRows = set()
        for row in sheet.iter_rows(values_only=True):
            uniqueRows.add(tuple(row))
        return len(uniqueRows), None
    except Exception as e:
        return None, e

if __name__ == "__main__":
    main()
